### ISSUE TITLE
Terrain rendering issues with TAA/FSR (flickering)

### ISSUE THREAD

**Issue by DevLewa** [TokisanGames/Terrain3D#302]
### Terrain3D version

v0.9.0

### System information

Godot v4.2.1 - Windows 10, Forward+/Vulkan GTX 1060 6GB

### Is the issue reproducable in the demo?

Not applicable

### Issue description

Jittering/flickering can be seen/observed on the Terrain if TAA or FSR 2.0 is enabled. (With TAA it is less noticeable, with FSR its more pronounced)

Here a demonstration:
https://www.youtube.com/watch?v=mKjPz8nHW2U

At first thought that this is a general Godot issue, howerver in the video only the terrain experiences flickering. (Other static geometry looks fine.)
I assume that this is due to the snapping/reposition behaviour of the 3D terrain renderer given that this flicker happens in a fixed intervall while moving. At least FSR to my knowledge requires a velocity buffer to reconstruct the image. The repositioning of the terrain might interfere with the FSR implementation casing incorrect reconstruction.

Not entirely sure how the velocity tracking in Godot is implemented. If we store the previous transform of the last frame (to calculate the delta for the velocity buffer each frame) one fix would be to artifically override the last transform state in case of a reposition to trick the image reconstruction on applying the correct velocity.

Can't test this with the demo project as of right now. (Having issues with godot not wanting to enable the addon on new Projects. Will update this once i figure out what is going on.)

**Comment by TokisanGames** [comment]
Thanks for the report. I haven't used either yet. Is it the mesh or the textures flickering? Show it in wireframe mode, and checkerboard and grey debug views. Your trees flicker just as much.

**Comment by DevLewa** [comment]
Here a video with the default terrain grid texture and with enabled collision mesh:
https://www.youtube.com/watch?v=JJZRwbQwQiI

The trees flicker constantly (even stationary) as they have a material with alpha transparency + alpha clipping enabled.  (without FSR is already really aliased + i'ts a really low resolution scale for demo purposes so it looks especially bad here.) However the terrain flickering shows a different behaviour. (It flickers only while moving in fixed interfals.)

You can also see that the grey grid texture on the cubes to the left are always stable while the floor is constantly jittering.
Hence my assumption that this is related to the velocity buffer issue.

**Comment by InitialCon** [comment]
It's very likely that this is a motion vector problem. The height of the vertex is determined by its world position so when the terrain moves with the camera the position of the vertex will move and this will not write to the motion vector buffer. This is a problem since the fragment sample position moves too. Since TAA and FSR2 expect pixel colours to remain constant with a constant velocity it's reprojecting the previous sample from before the camera moved causing the blurring/flickering. I'm not sure if it's possible to fix this without outputting the velocity in the shader which is not currently supported. My TAA understanding isn't the best so I may be incorrect here or get some terminology wrong.

**Comment by TokisanGames** [comment]
I just tried FSR 1.0 and 2.2 in Godot 4.2.1 with the Terrain3D demo and they display perfectly fine on my NVidia 3070. 

However when trying to load a project or a new terrain on my system it reports errors:
```
ERROR: Condition "p_job.uavMip[i] >= mip_slice_rids.size()" is true. Returning: FFX_ERROR_INVALID_ARGUMENT
   at: execute_gpu_job_compute_rd (servers/rendering/renderer_rd/effects/fsr2.cpp:418)
```

It also repeats these when moving the mouse around, and the decal mouse cursor doesn't show.

I just helped a user on discord who was having Godot crash with FSR 2.2 enabled on his NVidia 3080 while adding Terrain3D, with these errors:
```
ERROR: Width must be equal or greater than 1 for all textures
   at: (drivers/vulkan/rendering_device_vulkan.cpp:1722)
ERROR: Condition "texture.is_null()" is true. Returning: FFX_ERROR_BACKEND_API_ERROR
   at: create_resource_rd (servers/rendering/renderer_rd/effects/fsr2.cpp:240)
```

He probably needs to upgrade his drivers. 

So, in conclusion, FSR 2.2 in Godot is not ready for production use. Even if it works for you, you have no idea if it will be stable on end user machines. We could possibly put some test conditions in Terrain3D to avoid problems if we know the triggers. However, the back end in the engine clearly has bugs which need to be identified, reported, and fixed on their end.

My recommendation is to not use it until it's much more stable.

**Comment by TokisanGames** [comment]
Regarding TAA: I do see occasional flickering in the demo in some sections of the rock texture, not all. It's minor. If I zoom up the textures so they're pixelated and set the filtering to nearest, it triggers TAA artifacts much more. 

In Op's case, it looks like all of the textures are chosen specifically to conflict with TAA. If I put the texture on a mesh it looks fine. As @InitialCon wrote, the difference is that the terrain mesh moves around. Put the viewport in wireframe mode and you can see that when you laterally move the camera the mesh moves constantly. There's, very little TAA artifacts if the camera moves vertically only.

So TAA is likely not going to work. FSR seems fine in the demo, but maybe not with other textures, and is unstable anyway.

For now use MSAA or FXAA.

**Comment by mrjustaguy** [comment]
I've observed that the motion vectors just shoot all over the place with the terrain (and nothing else), so it's a MV issue that's breaking TAA and FSR

**Comment by TokisanGames** [comment]
[Wrobot's latest jungle demo](https://wrobot.itch.io/jungledemo) uses TAA in 4.3 with Terrain3D and it looks great. Terrain textures are stable. Very few artifacts in lighting only, unrelated to Terrain3D. 

It also tests fxaa, msaa, and FSR. No issues on my 3070.

**Comment by mrjustaguy** [comment]
I'm looking in the Motion Vector Debug view in 4.3 on an AMD Radeon RX 6600, and I'm getting this: 
![image](https://github.com/user-attachments/assets/7520d5d4-5a07-4d75-b1f9-1a9e3f769cd8)

The MVs randomly flicker like that, and it's only happening for the terrain, while other things provide proper motion vectors such as that box

Do also note, that at times the terrain seems to be providing proper MVs too, it just keeps going completely crazy every handful of frames

**Comment by mrjustaguy** [comment]
Interesting fact I've just discovered, the Motion vectors go crazy differently for different LODs, if you scroll from side to side when looking from top down on a flat terrain and look at the MV view, you'll be able to see the different LODs represented in the Motion vectors

![image](https://github.com/user-attachments/assets/89b7dde0-d773-486d-93f7-f9066e28c892)

![image](https://github.com/user-attachments/assets/c77a86e6-f5ca-456a-97ac-c9832a1cd232)

This combined with the fact this is showing up mostly during actual camera movement, and only when the Terrain is aware of the camera, I'd say the culprit of these wacky MVs is the Clip map poisition updating

**Comment by TokisanGames** [comment]
https://www.reddit.com/r/godot/comments/1bmcv9x/shader_motion_vectors_godot_421/

https://docs.godotengine.org/en/stable/contributing/development/core_and_modules/internal_rendering_architecture.html#temporal-antialiasing

> Motion vectors are generated on the GPU in the main material shader. This is done by running the vertex shader corresponding to the previous rendered frame (with the previous camera transform) in addition to the vertex shader for the current rendered frame, then storing the difference between them in a color buffer.

Perhaps we can look at the motion vector code in the main shader and invert it, or empty the color buffer when TAA is enabled.

Or perhaps we need a godot proposal to expose an API in the renderer to control the motion vectors to either:
* insert an inverse vector to neutralize the motion
* override the vector
* apply a mask to cull the terrain from motion vector calculation

**Comment by mrjustaguy** [comment]
https://github.com/godotengine/godot/pull/77523 may be the answer to this problem, as it'd disable motion vector generation, which if I understand it correctly, in the case of FSR would mean it'd Fall back to Derived Motion Vectors that wouldn't suffer from this problem

**Comment by TokisanGames** [comment]
77523, disabling motion vectors [does not help](https://github.com/godotengine/godot/pull/77523#issuecomment-2298414625) and makes things worse.

**Comment by mrjustaguy** [comment]
You tested it with TAA or FSR?

TAA to my knowledge doesn't derivate invalid MVs, and lack of MVs would mess things up

FSR does so if no MVs are provided, it'll derivate them and it should work fine.

if it was tested with TAA, please retest with FSR 2.2

**Comment by TokisanGames** [comment]
TAA and FSR issues are resolved by https://github.com/godotengine/godot/pull/105437 on my RTX 3070.

To test it you must:
1. Use Godot 4.5+
2. Use a version of godot-cpp that has RenderingServer::instance_teleport() or build godot-cpp with a [custom api json](https://docs.godotengine.org/en/latest/tutorials/scripting/gdextension/gdextension_cpp_example.html#building-the-c-bindings) generated from that engine binary. 
3. Add `RS->instance_teleport(mesh_array[instance]);` after this line: https://github.com/TokisanGames/Terrain3D/blob/6169f8c83a39e69f678e10e0ffa8eb5e65b41acf/src/terrain_3d_mesher.cpp#L392
4. [Build Terrain3D](https://terrain3d.readthedocs.io/en/latest/docs/building_from_source.html)

**Comment by TokisanGames** [comment]
https://github.com/godotengine/godot/pull/105437 was merged in. Instructions to use it are above. This ticket will be closed once these are done:
* godot-cpp has `instance_teleport()`
* We add `instance_teleport()` behind a 4.5 preprocessor directive.
* We remove the filter around physics interpolation since `main` is 4.4 only

**Comment by TokisanGames** [comment]
That's a godot/driver issue. It's not stable on all systems. Although we don't support it until the above is done.

**Comment by tavurth** [comment]
Is this also why I'm seeing a lot of jitter when motion blur is enabled?

### Motion Blur Enabled
https://github.com/user-attachments/assets/ec60a14d-5788-4ef0-a884-f0e73125dbc3

### Motion Blur Disabled
https://github.com/user-attachments/assets/9167b5dd-800f-40b7-be9b-23ffaef886e7

### Sample Project
https://drive.google.com/file/d/1XoonrFrWUcKBtb3qykLbiugjYTlAEhyo/view

**Comment by TokisanGames** [comment]
Disable TAA/FSR until you have done [the above steps](https://github.com/TokisanGames/Terrain3D/issues/302#issuecomment-2807536759), or they are built in. If you're not using those and you have added a motion blur shader, then you also need to add a way to reset your motion vectors after they are teleported, or disable the shader. Turn on the motion vector view and you'll see they are out of whack.

**Comment by tavurth** [comment]
> If you're not using those and you have added a motion blur shader, then you also need to add a way to reset your motion vectors after they are teleported

Yep not enabled any TAA, it's just from the composite shaders I guess.

Thanks for the links, I took a read and good to know there's a way to fix this before any release. 

I did try to run the sample above in the new Godot 4.5 beta-1, and the blur compute shaders seem to be all kinds of messed up at the moment so I'll leave it off for a while, and check again a little later 😅

**Comment by unfa** [comment]
I think I have hit this same issue when using Terrain3D and JFA Motion Blur:
https://github.com/user-attachments/assets/ca44a873-f4eb-4a3f-b6ea-a42b64a29944

It seems the PR required to allow fixing this in Godot has been merged months ago, but I still get the terrain flicker when any motion-vectors based effect is in use.

Here's TAA + FSR 2.2 at 25% render scale:
https://github.com/user-attachments/assets/063971dc-51bd-4648-8eda-e54c26bf0c75

**Comment by TokisanGames** [comment]
Motion blur is unrelated to this issue and will not be fixed by the solution here. I mentioned above what is required for motion blur shaders. If they are ever possible, they'll need to be addressed in the renderer. For now, don't use motion blur shaders. 

This issue is only for TAA/FSR, which are already fixed in Godot and accessible now. I listed above how to use it now, and what it will take to see it in main builds and close this issue.

**Comment by tavurth** [comment]
Super weird that since I swapped to my own shader based CLOD terrain system, I have no problems with using motion blur any more. 🤔

**Comment by tavurth** [comment]
I see you have marked all my replies as off topic? I have been trying to add something to the discussion here.

The above video report I posted happens with Terrain3D and FAA as well, but not with shader based CLOD.

It's a bug in this terrain library, not a bug affecting all Godot terrain systems.

**Comment by FunkyArmor** [comment]
> I see you have marked all my replies as off topic? I have been trying to add something to the discussion here.
> 
> The above video report I posted happens with Terrain3D and FAA as well, but not with shader based CLOD.
> 
> It's a bug in this terrain library, not a bug affecting all Godot terrain systems.

I suspected as much as well. I've been experiencing this issue for months now trying to have it resolved. Have you pin pointed the issue with this library?

**Comment by TokisanGames** [comment]
> I see you have marked all my replies as off topic?

@tavurth It's not personal. I marked replies about motion blur and CLOD terrain as off topic. This issue is about TAA/FSR on a clipmap terrain, which has been solved, but not merged.

> It's a bug in this terrain library, not a bug affecting all Godot terrain systems.

Not quite. The terrain meshes moves often, potentially every frame. This is inherent to how a clipmap terrain works. Godot introduced motion vectors when they added Physics Interpolation, TAA, FSR, without providing a facility to reset those motion vectors. Those features broke the terrain when used. They eventually provided a facility to reset those motion vectors for PI, then later TAA and FSR, but the latter haven't been propagated into Godot 4.4. All of this has been discussed right here, above.

>  I've been experiencing this issue for months now trying to have it resolved. 

@FunkyArmor PI/TAA/FSR have been fixed since April. You can use it today. I already laid out the [instructions](https://github.com/TokisanGames/Terrain3D/issues/302#issuecomment-2807536759) on how to incorporate the fix for your own game. If you don't want to build from source, then disable them until the fixes are built in. The schedule for that was listed [here](https://github.com/TokisanGames/Terrain3D/issues/302#issuecomment-2830916342).

It's possible those fixes to the renderer also fixed motion blur, but so far no motion blur users have tested it.

If you guys want to open a separate ticket about motion blur, fine, but that's a Godot issue and you should also open a linked issue on the Github repo, since we don't make the renderer. However, motion blur shaders aren't part of the engine, so the issue might belong with the creator. You'd have to talk with the creator and the rendering devs to figure out who is responsible for providing the facility for clearing the motion vectors. We can certainly make adjustments like issuing a signal when the terrain is moved. But without that facility, motion blur will not work with any clipmap terrain.

**Comment by DakotaPF** [comment]
hey guys!

I'm new to game dev and this is my first git post haha (sorry if my etiquette is off l0l) but i wanted to post here and say that after rebuilding Terrain3d according to the mentioned steps below (using 4.5.1 api), the JFA Motion Blur addon, "SphynxMotionBlurToolkit" is now functioning without any flickering issues! I think it was presumed that adding the `RS->instance_teleport(mesh_array[instance]);` line would result in a fix but i wanted to post the successful results for anyone else trouble shooting the issue :)

Also, I know it was mentioned that this isn't the right thread for motion blur related comments / issues but I couldn't find another thread.

x
dak

> TAA and FSR issues are resolved by [godotengine/godot#105437](https://github.com/godotengine/godot/pull/105437) on my RTX 3070.
> 
> To test it you must:
> 
> 1. Use Godot 4.5+
> 2. Use a version of godot-cpp that has RenderingServer::instance_teleport() or build godot-cpp with a [custom api json](https://docs.godotengine.org/en/latest/tutorials/scripting/gdextension/gdextension_cpp_example.html#building-the-c-bindings) generated from that engine binary.
> 3. Add `RS->instance_teleport(mesh_array[instance]);` after this line: 
>      
>        
>          [Terrain3D/src/terrain_3d_mesher.cpp](https://github.com/TokisanGames/Terrain3D/blob/6169f8c83a39e69f678e10e0ffa8eb5e65b41acf/src/terrain_3d_mesher.cpp#L392)
>        
>        
>             Line 392
>          in
>          [6169f8c](/TokisanGames/Terrain3D/commit/6169f8c83a39e69f678e10e0ffa8eb5e65b41acf)
>        
>      
>      
>        
>    
>            
>              
>               RS->instance_set_transform(mesh_array[instance], t);
> 4. [Build Terrain3D](https://terrain3d.readthedocs.io/en/latest/docs/building_from_source.html)

**Comment by TokisanGames** [comment]
Now that the engine has merged in the changes necessary for TAA, FSR, and motion blur shaders, #885 incorporates the necessary code change to use it.

@DakotaPF Thanks for the report.

To use Terrain3D with TAA, FSR2, or Motion Blur you must:
* [Build Terrain3D](https://terrain3d.readthedocs.io/en/latest/docs/building_from_source.html) from source using godot-cpp 4.5.
* Use the Godot 4.5+ engine binary.

Terrain3D currently builds against the 4.4 API, which works in later versions, but doesn't have the new function necessary for this. Once we offer builds against the 4.5 API, building from source won't be necessary.

We're not ready to abandon 4.4 yet, nor maintain two builds. Perhaps in 2026 I'll start a secondary branch to build both 4.4 and 4.5. In the meantime, you can build from source, or disable TAA/FSR/MB. With this change, you no longer need to change the code.

**Comment by TokisanGames** [comment]
`main` has been updated to godot-4.5, no longer requiring a custom build for temporal effects.
`1.1-godot4.4` has been created and will be maintained for a bit through the actual 1.1 release.
Godot binaries 4.4+ work with either, barring issues with 4.6.
